doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,653 @@
+write), not both. This call will not be implemented for QUIC (e.g. always
+returns `SSL_NOTHING`) and `SSL_want_net_read` and `SSL_want_net_write` will be
+used instead.
+
+TBD: Should these be implemented for non-QUIC SSL objects?
+
+#### `SSL_set_initial_peer_addr`, `SSL_get_initial_peer_addr`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | No        | CS            |
+
+`SSL_set_initial_peer_addr` sets the initial L4 UDP peer address for an outgoing
+QUIC connection.
+
+The initial peer address may be autodetected if no peer address has already been
+set explicitly and the QUIC connection SSL object is provided with a
+`BIO_s_dgram` with a peer set.
+
+`SSL_set_initial_peer_addr` cannot be called after a connection is established.
+
+### Future APIs
+
+A custom poller interface may be provided in the future. For more information,
+see the QUIC I/O Architecture design document.
+
+BIO Objects
+-----------
+
+### Existing APIs
+
+#### `BIO_s_connect`, `BIO_new_ssl_connect`, `BIO_set_conn_hostname`
+
+We are aiming to support use of the existing `BIO_new_ssl_connect` API with only
+minimal changes. This will require internal changes to `BIO_s_connect`, which
+should automatically detect when it is being used with a QUIC `SSL_CTX` and act
+accordingly.
+
+#### `BIO_new_bio_pair`
+
+Unsuitable for use with QUIC on the network side; instead, applications can
+make use of the new `BIO_s_dgram_pair` which provides equivalent functionality
+with datagram semantics.
+
+#### Interactions with `BIO_f_buffer`
+
+Existing applications sometimes combine a network socket BIO with a
+`BIO_f_buffer`. This is problematic because the datagram semantics of writes are
+not preserved, therefore the BIO provided to libssl is, as provided, unusable
+for the purposes of implementing QUIC. Moreover, output buffering is not a
+relevant or desirable performance optimisation for the transmission of UDP
+datagrams and will actually undermine QUIC performance by causing incorrect
+calculation of ACK delays and consequently inaccurate RTT calculation.
+
+Options:
+
+  - Require applications to be changed to not use QUIC with a `BIO_f_buffer`.
+  - Detect when a `BIO_f_buffer` is part of a BIO stack and bypass it
+    (yucky and surprising).
+
+#### MTU Signalling
+
+**See also:**
+[BIO_s_dgram_pair(3)](https://www.openssl.org/docs/manmaster/man3/BIO_s_dgram_pair.html)
+
+`BIO_dgram_get_mtu` (`BIO_CTRL_DGRAM_GET_MTU`) and `BIO_dgram_set_mtu`
+(`BIO_CTRL_DGRAM_SET_MTU`) already exist for `BIO_s_dgram` and are implemented
+on a `BIO_s_dgram_pair` to allow the MTU to be determined and configured. One
+side of a pair can configure the MTU to allow the other side to detect it.
+
+`BIO_s_dgram` also has pre-existing support for getting the correct MTU value
+from the OS using `BIO_CTRL_DGRAM_QUERY_MTU`.
```

Comment:
There's also a `BIO_CTRL_DGRAM_GET_FALLBACK_MTU`. How does that fit in?

Reply:
That returns a static value which should work in all circumstances for a given address family. Not sure it really comes into play for QUIC, since QUIC specifies its own minimum MTU (1200 bytes datagram payload size) and any increase above that will be based on some kind of signalling or negotiation.